analyses/doublet-detection/scripts/00_benchmark-methods.R

```diff
@@ -0,0 +1,251 @@
+
+
+# Define paths and set seed ------------------
+repository_base <- rprojroot::find_root(rprojroot::is_git_root)
+data_dir <- file.path(repository_base, "data", "current")
+module_base <- file.path(repository_base, "analyses", "doublet-detection")
+
+# directory to save doublet detection results
+output_dir <- file.path(module_base, "results", "benchmark-results")
+fs::dir_create(output_dir)
+
+# directory to save temporary files
+scratch_dir <- file.path(module_base, "scratch", "benchmark-sces")
+fs::dir_create(scratch_dir)
+
+# output files
+benchmark_runtime_tsv <- file.path(output_dir, "benchmark_runtimes.tsv")
+benchmark_results_tsv <- file.path(output_dir, "benchmark_results.tsv")
+
+set.seed(11)
+
+# Read in all sample metadata -----------------
+metadata_files <- list.files(
+  data_dir,
+  pattern = "single_cell_metadata.tsv",
+  full.names = TRUE,
+  recursive = TRUE
+)
+names(metadata_files) <- stringr::str_extract(metadata_files, "SCPCP\\d{6}")
+
+# Exclude multiplexed, for now
+metadata_files <- metadata_files[names(metadata_files) != "SCPCP000009"]
+
+# data frame of all library diagnoses and processed cell counts
+sample_df <- metadata_files |>
+  purrr::imap(
+    \(file, project_id) {
+      readr::read_tsv(file) |>
+        dplyr::select(scpca_sample_id,
+                      scpca_library_id,
+                      diagnosis,
+                      processed_cells) |>
+        dplyr::mutate(project_id = project_id) |>
+        dplyr::select(project_id, everything())
+
+    }
+  ) |>
+  dplyr::bind_rows()
+
+# Identify libraries to use ------------------------
+
+# We'll keep 7 libraries per project (or all libraries, if < 10 total)
+# across a range of library sizes - rank each and select bottom 2, middle 3, top 2
+
+min_cells <- 50
+
+libraries_to_benchmark <- sample_df |>
+  # keep only libraries with >= min_cells cells
+  dplyr::filter(processed_cells >= min_cells) |>
+  dplyr::group_by(project_id) |>
+  dplyr::mutate(n_samples = dplyr::n()) |>
+  dplyr::arrange(processed_cells) |>
+  dplyr::slice(c(
+    1:2, # bottom 2
+    (floor(n_samples/2) - 1):(floor(n_samples/2) + 1), # roughly middle 3
+    (n_samples - 1):n_samples # top 2
+  )) |>
+  dplyr::ungroup() |>
+  # remove duplicates for projects with <10 libraries
+  dplyr::distinct() |>
+  dplyr::select(-n_samples) |>
+  # add column with file path
+  dplyr::mutate(file_path = file.path(
+    data_dir, project_id, scpca_sample_id, glue::glue("{scpca_library_id}_processed.rds")
+  ))
+
+# Print samples as a comma-sep list for data download
+#print(paste0(libraries_to_benchmark$scpca_sample_id, collapse = ","))
+
+# Run doublet detection methods and export results ------------------------------------
+
+sce_files <- libraries_to_benchmark$file_path
+names(sce_files) <- libraries_to_benchmark$scpca_library_id
+
+doublet_output <- sce_files|>
+  purrr::imap(benchmark_doublets, scratch_dir)
+
+runtime_df <- doublet_output |>
+  purrr::map(\(x) x$runtimes) |>
```

Comment:
purrr has you covered here! You want to just select the same element? It's got you!
```suggestion
  purrr::map("runtimes") |>
```